Integration merlion issue2377

rfsiten opened the pull request:
1) Modify sql_query.py to support to define customized return columns and dtype of ml_handlers;
2) [issue2377] Merlion integrated, forecaster: default, sarima, prophet, mses, detector: default, isolation forest, windstats, prophet;
3) Corresponding test cases added, /mindsdb/tests/unit/test_merlion_handler.

rfsiten commented:
hey, it's the initial version of merlion-integration @paxcema  @ZoranPandovski, please review and feel free to give me your advice.
Thank you!
 [issues2377](https://github.com/mindsdb/mindsdb/issues/2377)

rfsiten commented:
hey, it's the initial version of merlion-integration @paxcema  @ZoranPandovski, please review and feel free to give me your advice.
Thank you!
 [issues2377](https://github.com/mindsdb/mindsdb/issues/2377)

rfsiten commented:
some examples:

```
# forecast default

CREATE PREDICTOR merlion.test_default_forecaster
FROM test_mindsdb_proxy
(select t, H1 from m4_csv where train = 1) 
PREDICT H1;

select p.t, p.H1 real, t.H1, t.H1__upper, t.H1__lower
from merlion.test_default_forecaster t
inner join test_mindsdb_proxy.m4_csv p on t.t = p.t
where p.train = 0;

# forecast sarima

CREATE PREDICTOR merlion.test_sarima_forecaster
FROM test_mindsdb_proxy
(select t, H1 from m4_csv where train = 1) PREDICT H1
USING model_type='sarima'
horizon 50;

select p.t, p.H1 real, t.H1, t.H1__upper, t.H1__lower
from merlion.test_sarima_forecaster t
inner join test_mindsdb_proxy.m4_csv p on t.t = p.t
where p.train = 0;

# forecast prophet

CREATE PREDICTOR merlion.test_prophet_forecaster
FROM test_mindsdb_proxy
(select t, H1 from m4_csv where train = 1) PREDICT H1
USING task='forecast', model_type='prophet'
horizon 50;

select p.t, p.H1 real, t.H1, t.H1__upper, t.H1__lower
from merlion.test_prophet_forecaster t
inner join test_mindsdb_proxy.m4_csv p on t.t = p.t
where p.train = 0;


# forecast mses

CREATE PREDICTOR merlion.test_mses_forecastor
FROM test_mindsdb_proxy
(select t, H1 from m4_csv where train = 1) PREDICT H1
USING model_type='mses'
horizon 50;

select p.t, p.H1 real, t.H1, t.H1__upper, t.H1__lower
from merlion.test_mses_forecastor t
inner join test_mindsdb_proxy.m4_csv p on t.t = p.t
where p.train = 0;


# detect default

CREATE PREDICTOR merlion.default_anomaly_detector
FROM test_mindsdb_proxy
(SELECT t, val FROM mindsdb_test.nba_anomaly_csv where train = 1) PREDICT val
using task='detector';


select t, val__anomaly_score from(
select p.t, d.val__anomaly_score
from merlion.default_anomaly_detector d
inner join test_mindsdb_proxy.nba_anomaly_csv p on d.t = p.t and d.val = p.val
where p.train = 0) m
where val__anomaly_score > 0

# detect isolation

CREATE PREDICTOR merlion.isolation_anomaly_detector
FROM test_mindsdb_proxy
(SELECT t, val FROM mindsdb_test.nba_anomaly_csv where train = 1) PREDICT val
using task='detector', model_type='isolation';


select t, val__anomaly_score from(
select p.t, d.val__anomaly_score
from merlion.isolation_anomaly_detector d
inner join test_mindsdb_proxy.nba_anomaly_csv p on d.t = p.t and d.val = p.val
where p.train = 0) m
where val__anomaly_score > 0;

# detect windstats

CREATE PREDICTOR merlion.windstats_anomaly_detector
FROM test_mindsdb_proxy
(SELECT t, val FROM mindsdb_test.nba_anomaly_csv where train = 1) PREDICT val
using task='detector', model_type='windstats';


select t, val__anomaly_score from(
select p.t, d.val__anomaly_score
from merlion.windstats_anomaly_detector d
inner join test_mindsdb_proxy.nba_anomaly_csv p on d.t = p.t and d.val = p.val
where p.train = 0) m
where val__anomaly_score > 0;

# detect prophet

CREATE PREDICTOR merlion.prophet_anomaly_detector
FROM test_mindsdb_proxy
(SELECT t, val FROM mindsdb_test.nba_anomaly_csv where train = 1) PREDICT val
using task='detector', model_type='prophet';

select t, val__anomaly_score from(
select p.t, d.val__anomaly_score
from merlion.prophet_anomaly_detector d
inner join test_mindsdb_proxy.nba_anomaly_csv p on d.t = p.t and d.val = p.val
where p.train = 0) m
where val__anomaly_score > 0;
```

ZoranPandovski commented on mindsdb/integrations/handlers/merlion_handler/__about__.py:
```diff
@@ -0,0 +1,9 @@
+__package_name__ = 'mindsdb_merlion_handler'
+__version__ = '0.0.1'
+__description__ = "MindsDB handler for Merlion"
+__author__ = 'MindsDB Inc'
```
Please add yourself as an author :raised_hands:

ZoranPandovski commented on mindsdb/integrations/handlers/merlion_handler/adapters.py:
```diff
@@ -0,0 +1,162 @@
+                                                               target_seq_index=self.TARGET_SEQ_INDEX))
+
+
+class SarimaForecasterAdapter(BaseMerlionForecastAdapter):
```
Maybe we can add doc string for each adapter and reference to the merlion docs for more info?

ZoranPandovski commented on tests/unit/test_merlion_handler.py:
```diff
@@ -0,0 +1,202 @@
+class TestMerlion(BaseExecutorTest):
+    def get_m4_abs_path(self):
+        dir_path = os.path.dirname(os.path.realpath(__file__))
+        return os.path.join(dir_path, "m4_test_merlion.csv")
```
Should we keep the files somwhere else and reference them from there?

rfsiten replied:
All the local files have been replaced with github url.

ZoranPandovski replied:
Thanks

rfsiten commented:
hey, dear @paxcema @ZoranPandovski , I've try my best to find why my pr's check failed as below:
===============
test MySQL API
===============

Warning: Permanently added '3.220.66.106' (ECDSA) to the list of known hosts.
Load key "/home/runner/.ssh/db_machine": invalid format
ubuntu@3.220.66.106: Permission denied (publickey).

In fact, I haven't modified any mysql related code. Could you provide me some pieces of advice?

ZoranPandovski commented:
@rfsiten Actually you can't fix that since it is one of our environments that needs to be started and you don't have access. We need to improve this tests, or for now don't run this if the PR comes from `forked` repo. @paxcema Is off today so I guess he will do another review on Monday

rfsiten commented:
> fied any mysql related code. Could you provide me some pieces of advice?



> @rfsiten Actually you can't fix that since it is one of our environments that needs to be started and you don't have access. We need to improve this tests, or for now don't run this if the PR comes from `forked` repo. @paxcema Is off today so I guess he will do another review on Monday

Got it, thank you!

paxcema commented on mindsdb/integrations/handlers/merlion_handler/merlion_handler.py:
```diff
@@ -0,0 +1,223 @@
+        if task_enum == TaskType.forecast:
+            model_args[MerlionArguments.max_forecast_steps.value] = horizon
+        adapter: BaseMerlionForecastAdapter = adapter_class(**model_args)
+        print("Training model, args: " + json.dumps(args))
```
Best to redirect all `print`s to the mindsdb logger

rfsiten replied:
> Best to redirect all `print`s to the mindsdb logger

all "prints" have been replaced by "mindsdb.log"

paxcema commented on mindsdb/integrations/handlers/merlion_handler/requirements.txt:
```diff
@@ -0,0 +1,5 @@
+mindsdb_sql >= 0.4.0
+pandas <=1.4.3
+
+salesforce-merlion
```
I would recommend capping this on both lower and upper limits to avoid situations where the handler may stop working after an upstream release.

rfsiten replied:
> I would recommend capping this on both lower and upper limits to avoid situations where the handler may stop working after an upstream release.

Lower and upper bound added for salesforce-merlion [1.2.0 ~ 1.3.1]

paxcema reviewed (changes requested):
Overall this looks great! Amazing effort @rfsiten 🚀 

I have pinged @ea-rus to discuss the proposed changes to the datanode and base ML engine. Maybe there is an alternative way of accomplishing the same without modifications to those classes.

rfsiten commented:
> Overall this looks great! Amazing effort @rfsiten 🚀
> 
> I have pinged @ea-rus to discuss the proposed changes to the datanode and base ML engine. Maybe there is an alternative way of accomplishing the same without modifications to those classes.

yeah, dear @paxcema , it's the simplest way I can find now, but I think you have the overall viewsight more broadly than I have, welcome your new advice, I am looking forward to it 🚀

ea-rus commented on mindsdb/api/mysql/mysql_proxy/classes/sql_query.py:
```diff
@@ -1161,6 +1161,16 @@ def execute_step(self, step, steps_data):
+                            columns_collection.add(table_name, (col, col))
+                    else:
+                        base_ml_engine_type_dict = {}
+
```
Do I understand right: it was done to add columns to result from predictor?  But you can just return this columns from predict method and they will be added to result

rfsiten replied:
If I only return these columns in prediction method but without the modification in sql_query.py, exception will be raised in the phase of sql-parsing.

ea-rus replied:
What exception do you have in sql_query.py without this block?

I made a little test: removed all your changes from sql_query, added coumn at the end of predict method in handler.
I see no error in sql_query, this column is passed to result and test are also working  
![image](https://user-images.githubusercontent.com/8502631/198241565-5763be19-1efe-4ecf-98d0-f8b33ff93e08.png)
![image](https://user-images.githubusercontent.com/8502631/198241619-73b281a9-6ac2-4151-8c55-01bd03e69fdd.png)

rfsiten replied:
hey bros, I nearly can't believe my eyeball. I deleted all my adding code in sql_query and other related codes, run the end-to-end test, passed! run the test cases, passed, Again!!!

My dream of being super programmer ended ... however, I am very clearly remember that it would crash without that codes.

Anyway,  forget that, I've refined the codes, please have a check. Forgive me ... Thank you!

ea-rus commented on mindsdb/integrations/handlers/merlion_handler/requirements.txt:
```diff
@@ -0,0 +1,5 @@
+mindsdb_sql >= 0.4.0
+pandas <=1.4.3
+
+1.2.0 <= salesforce-merlion <= 1.3.1
```
Is it right format? My pip shows exception

rfsiten replied:
fixed

ea-rus commented on mindsdb/integrations/handlers/merlion_handler/merlion_handler.py:
```diff
@@ -0,0 +1,223 @@
+
+# remove ` from column, the result is (mapping, back_mapping)
+# this function is a little strange, but I think it's important to keep
+# the raw column names, maybe ` is a trick while parsing sql? maybe ...
```
I can't repeat the case when column contents '`'. How I can do this.
And I think pandas can have '`' in column name and it shouldn't brake anything

rfsiten replied:
I found '`' when the input columns has the column name 'id'.

ea-rus replied:
Can't repeat with id too..
If this is not hard, can you send me little script with which I can catch '\`' in handler?  
I think this problem (if it is) should be solved outside of ml handler

rfsiten replied:
I remember the problem appeared while 'id' column existed. I can't repeat it just now. I guess there maybe exist an upgrading in mindsdb-sql or sql_query.py. But it's a good news, it means the work of developing a handle has been simplified. I will modify my code and push soon.

ea-rus commented on mindsdb/integrations/handlers/merlion_handler/merlion_handler.py:
```diff
@@ -0,0 +1,224 @@
+        task = args.get(self.ARG_USING_TASK, TaskType.forecast.name)
+        model_type = args.get(self.ARG_USING_MODEL_TYPE, self.DEFAULT_MODEL_TYPE)
+        time_column = args.get(self.ARG_USING_TIME_COLUMN, None)
+        horizon = args.get(self.ARG_HORIZON, self.DEFAULT_MAX_PREDICT_STEP)
```
horizon and window are defined in BaseMLEngineExec. So we can use these names as convectional, without special constant. Just args.get('horizon'...)

rfsiten replied:
Because I used this string twice， so I think maybe use a constant will be easier to maintain the code.
No problem, I've use string directly to replace the constant.

ea-rus commented on mindsdb/integrations/handlers/merlion_handler/merlion_handler.py:
```diff
@@ -0,0 +1,224 @@
+        log.info("Model and args saved.")
+
+        # back mapping column name
+        df.rename(columns=column_name_back_mapping, inplace=True)
```
looks like this line is not used

rfsiten replied:
I supposed maybe the df will be used by the code out of this scope.
Forget it, I have deleted these two lines.

ea-rus commented on tests/unit/test_merlion_handler.py:
```diff
@@ -0,0 +1,220 @@
+import requests
+
+
+class TestMerlion(BaseExecutorTest):
```
Tests are working for you? For me they fail by some reason (need to debug to figure out).

rfsiten replied:
hey， please use the command "PYTHONPATH=./ pytest tests/unit/test_merlion_handler.py", at first, please change current path to ./mindsdb/. Here I should apologize that I‘ve written the mistaken command in comment as "PYTHONPATH=./ pytest tests/unit/test_ml_handlers.py"， the right command should be "PYTHONPATH=./ pytest tests/unit/test_merlion_handler.py".

rfsiten replied:
All the four test cases can pass in this way.

rfsiten replied:
Problem was brought to the pr in the latest comits... I will fix it, sorry...

rfsiten replied:
Fixed

ea-rus replied:
Yes, now tests are working. 
Thanks!

rfsiten replied:
my pleasure

ea-rus commented:
@rfsiten, also integration tests failed (in github actions). I don't see in changes what could be the cause of error. Maybe you changes are based on old staging branch and it needs to be updated.

rfsiten commented:
> @rfsiten, also integration tests failed (in github actions). I don't see in changes what could be the cause of error. Maybe you changes are based on old staging branch and it needs to be updated.

hey, I've merged the latest version of staging, but problem still existed, maybe it's a problem of enviroment setting, do you have some ideas?

rfsiten commented:
> @rfsiten Actually you can't fix that since it is one of our environments that needs to be started and you don't have access. We need to improve this tests, or for now don't run this if the PR comes from `forked` repo. @paxcema Is off today so I guess he will do another review on Monday

@ea-rus hey, could above enlighten us?

rfsiten commented:
> @rfsiten, also integration tests failed (in github actions). I don't see in changes what could be the cause of error. Maybe you changes are based on old staging branch and it needs to be updated.

Today, I will try to rollback all my modification to see whether it could pass the test.

rfsiten commented:
close by mistaken... reopen it

rfsiten commented:
close by mistaken... reopen it

rfsiten commented:
hey, @ea-rus @paxcema , now the pr nearly include no change but also failed ... could you find the problem?

rfsiten commented:
I've push all my commits to the branch again.

paxcema reviewed (changes requested):
Okay, I looked into this more carefully, here's a few thoughts:

paxcema commented on mindsdb/integrations/handlers/merlion_handler/adapters.py:
```diff
@@ -0,0 +1,182 @@
+    def to_bytes(self):
+        return self.model.to_bytes()
+
+    def from_bytes(self, bytes):
```
`to_bytes` and `from_bytes` should be consistent in signature. My vote is for `return`ing `self.model` in `from_bytes()`

rfsiten replied:
ok, to_bytes will have the return signature bytes,

from_bytes has been renamed to initialize_model, I think this is what the function do.

paxcema commented on mindsdb/integrations/handlers/merlion_handler/adapters.py:
```diff
@@ -0,0 +1,182 @@
+        if err_ts is None:
+            err_df = None
+        else:
+            err_df = pred_ts.to_pd()
```
Why is `err_df` set to `pred_ts`? Seems a bit weird because it means joining a `df` with itself on line 77, right? Shouldn't it be `err_df = err_ts.to_pd()`?

rfsiten replied:
> Why is `err_df` set to `pred_ts`? Seems a bit weird because it means joining a `df` with itself on line 77, right? Shouldn't it be `err_df = err_ts.to_pd()`?

Here, I suppose sometimes the err_df will miss some lines, so I use the left join to keep the join result has the same lines as input.

rfsiten replied:
You are right, should be err_df = err_ts.to_pd()

paxcema commented on mindsdb/integrations/handlers/merlion_handler/merlion_handler.py:
```diff
@@ -0,0 +1,193 @@
+            df.index = pd.to_datetime(df[time_col])
+        if time_col is None:
+            raise Exception("can not find datetime column for time series")
+    df.drop(columns=[time_col], inplace=True)
```
Side-effects like this are best avoided. Can you modify this method to explicitly return a new dataframe on which you perform all changes (i.e. modifying index, `pop()`ing columns, etc.)?

rfsiten replied:
ok, I will return a fully new dataframe. You are right, it's not wise to modify on the input.

paxcema commented on mindsdb/integrations/handlers/merlion_handler/merlion_handler.py:
```diff
@@ -0,0 +1,193 @@
+    DEFAULT_MAX_PREDICT_STEP = 100
+    DEFAULT_PREDICT_BASE_WINDOW = 10
+
+    PERSISIT_MODEL_FILE_NAME = "merlion_model"
```
Small typo: `PERSISIT` -> `PERSIST`

rfsiten replied:
> Small typo: `PERSISIT` -> `PERSIST`

sorry , It's my fault...

paxcema commented on mindsdb/integrations/handlers/merlion_handler/merlion_handler.py:
```diff
@@ -0,0 +1,193 @@
+            raise Exception("missing required key in args: " + self.KWARGS_DF)
+        else:
+            column_sequence = list(df.columns.values)
+            sorted(column_sequence)
```
This isn't doing anything. The in-place variant is `list.sort()`. However, as mentioned above, better to avoid side effects, so I'd ask for:

```
column_sequence = sorted(list(df.columns.values))
```

paxcema commented on mindsdb/integrations/handlers/merlion_handler/merlion_handler.py:
```diff
@@ -0,0 +1,193 @@
+        time_column = args[self.ARG_USING_TIME_COLUMN]
+        target = args[self.ARG_TARGET]
+        horizon = args["horizon"]
+        # window = args[self.ARG_WINDOW]
```
If `window` is not needed, then you can remove this comment 👍

rfsiten replied:
fixed

paxcema commented on mindsdb/integrations/handlers/merlion_handler/setup.py:
```diff
@@ -0,0 +1,25 @@
+from setuptools import setup, find_packages
```
Just realized this now, but you should the attributes below from `__about__`, because it is currently undefined.

rfsiten replied:
fixed

from .__about__ import __dict__ as about

paxcema commented on tests/unit/test_merlion_handler.py:
```diff
@@ -0,0 +1,224 @@
+        # test default
+        self.exec_train_and_forecast(mock_handler=mock_handler, model_name="default", using="")
+        # # test arima
+        # self.exec_train_and_predict(mock_handler=mock_handler, model_name="arima", using=f"USING model_type='arima'")
```
Curious as to why these are disabled? Presumably it's fast enough when fitting a single series, correct?

rfsiten replied:
ok, I will uncomment it.

paxcema replied:
I was just curious 😛 . Enable the ones you deem necessary for testing this handler thoroughly, and the rest can be removed (not necessary to have them as comments).

ea-rus commented:
I did run 'test_company_independent.py' test locally from this branch and it passed OK. In this PR I don't see changes that could brake this test. 
Looks like tests were not passed due some problem with tests and external repository, not because changes in this PR

rfsiten commented:
hey @ea-rus @paxcema the problems fixed, and I've modified the test cases for adapting the adding of 'project'.

rfsiten commented:
> hey @ea-rus @paxcema the problems fixed, and I've modified the test cases for adapting the adding of 'project'.

test cases of merlion handler passed

ea-rus commented:
Only one remark from my side:
Yesterday #3765 was merged and added these parameters to timeseries_settings 
```
        if statement.horizon is not None:
            problem_definition['horizon'] = int(statement.horizon)

        if statement.window is not None:
            problem_definition['window'] = int(statement.window)
```
You could get it from args['timeseries_settings'] in create method.

rfsiten commented:
> Only one remark from my side: Yesterday #3765 was merged and added these parameters to timeseries_settings
> 
> ```
>         if statement.horizon is not None:
>             problem_definition['horizon'] = int(statement.horizon)
> 
>         if statement.window is not None:
>             problem_definition['window'] = int(statement.window)
> ```
> 
> You could get it from args['timeseries_settings'] in create method.
> 
> Except of it I have no objection to merge

oh...  such a mysterious variable, I will use it to replace the code I write above.

rfsiten commented:
```
        if statement.horizon is not None:
            problem_definition['horizon'] = int(statement.horizon)

        if statement.window is not None:
            problem_definition['window'] = int(statement.window)
```
Code Above removed, and now use timeseries_settings.horizon/window.

Meanwhile, I've found the arguments in "args" variable also changed, now they have being moved to using.xxx, so I've modified my code to adapt the changes.

test_merlion_handler passed, please check and merge, thanks a lot.

@ea-rus @paxcema

paxcema reviewed (approved):
Looks good to me – apologies for the silent upstream changes, and thanks for your patience 🙏

rfsiten commented:
> Looks good to me – apologies for the silent upstream changes, and thanks for your patience pray

Never mind, we are very nearly to accomplish despite we've experienced such a long run. 
Thank you for your patient to review my code, too.

ZoranPandovski commented:
Great work @rfsiten thank you. @paxcema , @ea-rus This is ready for merge right?
